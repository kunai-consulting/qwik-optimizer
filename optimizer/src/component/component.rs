use crate::component::Language;
use crate::component::*;
use crate::segment::Segment;
use oxc_allocator::{Allocator, Box as OxcBox, CloneIn, IntoIn, Vec as OxcVec};
use oxc_ast::ast::*;
use oxc_ast::*;
use oxc_codegen::{Codegen, CodegenOptions};
use oxc_minifier::*;
use oxc_span::{SourceType, SPAN};
use serde::Serialize;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize)]
pub struct QrlComponent {
    pub id: Id,
    pub language: Language,
    pub code: String,
    pub qrl: Qrl,
}

impl QrlComponent {
    pub(crate) fn new(
        source_info: &SourceInfo,
        id: Id,
        exported_expression: Expression<'_>,
        imports: Vec<Import>,
        minify: bool,
        qrl_type: QrlType,
    ) -> QrlComponent {
        let language = source_info.language.clone();
        let qrl = Qrl::new(&id.local_file_name, &id.symbol_name, qrl_type);

        let source_type: SourceType = language.into();

        let code = Self::gen(
            &id,
            exported_expression,
            imports,
            minify,
            &source_type,
            &Allocator::default(),
        );
        QrlComponent {
            id,
            language: source_info.language.clone(),
            code,
            qrl,
        }
    }

    fn gen(
        id: &Id,
        exported_expression: Expression<'_>,
        imports: Vec<Import>,
        minify: bool,
        source_type: &SourceType,
        allocator: &Allocator,
    ) -> String {
        let name = &id.symbol_name;

        let ast_builder = AstBuilder::new(allocator);

        let id = OxcBox::new_in(ast_builder.binding_identifier(SPAN, name), allocator);
        let bind_pat = ast_builder.binding_pattern(
            BindingPatternKind::BindingIdentifier(id),
            None::<OxcBox<'_, TSTypeAnnotation<'_>>>,
            false,
        );
        let mut var_declarator = OxcVec::new_in(allocator);

        var_declarator.push(ast_builder.variable_declarator(
            SPAN,
            VariableDeclarationKind::Const,
            bind_pat,
            Some(exported_expression),
            false,
        ));

        let decl = ast_builder.variable_declaration(
            SPAN,
            VariableDeclarationKind::Const,
            var_declarator,
            false,
        );
        let decl = OxcBox::new_in(decl, allocator);
        let decl = Declaration::VariableDeclaration(decl);
        let export = ast_builder.export_named_declaration(
            SPAN,
            Some(decl),
            OxcVec::new_in(allocator),
            None,
            ImportOrExportKind::Value,
            None::<OxcBox<WithClause>>,
        );
        let export = Statement::ExportNamedDeclaration(OxcBox::new_in(export, allocator));

        let imports = imports.iter().map(|import| {
            let statement: Statement = import.clone().into_in(allocator);
            statement
        });

        let mut body = ast_builder.vec_from_iter(imports);

        body.push(export);

        let ast_builder = AstBuilder::new(allocator);

        let mut new_pgm = ast_builder.program(
            SPAN,
            *source_type,
            "",
            OxcVec::new_in(allocator),
            None,
            OxcVec::new_in(allocator),
            body,
        );

        let codegen = Codegen::new();
        let codegen_options = CodegenOptions {
            annotation_comments: true,
            minify: minify,
            ..Default::default()
        };

        if minify {
            let ops = MinifierOptions {
                compress: Some(CompressOptions::default()),
                mangle: None,
                // mangle: Some(MangleOptions::default()),
            };
            let minifier = Minifier::new(ops);
            let ret = minifier.build(allocator, &mut new_pgm);
            let sym_tab = ret.symbol_table;

            codegen
                .with_options(codegen_options)
                .with_symbol_table(sym_tab)
                .build(&new_pgm)
                .code
        } else {
            codegen.with_options(codegen_options).build(&new_pgm).code
        }
    }

    /// Create a QrlComponent from an `Expression`.
    pub(crate) fn from_expression(
        expr: Expression<'_>,
        imports: Vec<Import>,
        segments: &Vec<Segment>,
        target: &Target,
        scope: &Option<String>,
        source_info: &SourceInfo,
        minify: bool,
    ) -> QrlComponent {
        let qrl_type: QrlType = segments
            .last()
            .iter()
            .flat_map(|segment| segment.qrl_type())
            .last()
            .unwrap(); // TODO Clean this up.

        let id = Id::new(source_info, segments, target, scope);

        QrlComponent::new(source_info, id, expr, imports, minify, qrl_type)
    }

    pub(crate) fn from_call_expression_argument(
        arg: &Argument,
        imports: Vec<Import>,
        segments: &Vec<Segment>,
        target: &Target,
        scope: &Option<String>,
        source_info: &SourceInfo,
        minify: bool,
        allocator: &Allocator,
    ) -> QrlComponent {
        let init = arg.clone_in(allocator).into_expression();
        Self::from_expression(init, imports, segments, target, scope, source_info, minify)
    }
}
