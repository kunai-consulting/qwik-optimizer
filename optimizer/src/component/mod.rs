mod component;
mod id;
mod language;
mod qrl;
mod shared;
mod source_info;

// -- Flatten
pub(crate) use component::*;
pub(crate) use id::*;
pub(crate) use language::*;
pub(crate) use qrl::*;
pub(crate) use shared::*;
pub(crate) use source_info::*;
