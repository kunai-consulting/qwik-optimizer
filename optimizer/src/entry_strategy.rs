use crate::segment::Segment;
use serde::{Deserialize, Serialize};

const ENTRY_SEGMENTS: &str = "entry_segments";

// EntryStrategies
#[derive(Debug, Serialize, Copy, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum EntryStrategy {
    Inline,
    Hoist,
    Single,
    Hook,
    Segment,
    Component,
    Smart,
}

pub trait EntryPolicy: Send + Sync {
    fn get_entry_for_sym(&self, context: &[String], segment: &Segment) -> Option<String>;
}

#[derive(Default, Clone)]
pub struct InlineStrategy;

impl EntryPolicy for InlineStrategy {
    fn get_entry_for_sym(&self, _context: &[String], _segment: &Segment) -> Option<String> {
        Some(ENTRY_SEGMENTS.to_string())
    }
}

#[derive(Clone)]
pub struct SingleStrategy {}

impl SingleStrategy {
    pub const fn new() -> Self {
        Self {}
    }
}

impl EntryPolicy for SingleStrategy {
    fn get_entry_for_sym(&self, _context: &[String], _segment: &Segment) -> Option<String> {
        Some(ENTRY_SEGMENTS.to_string())
    }
}

#[derive(Clone)]
pub struct PerSegmentStrategy {}

impl PerSegmentStrategy {
    pub const fn new() -> Self {
        Self {}
    }
}

impl EntryPolicy for PerSegmentStrategy {
    fn get_entry_for_sym(&self, _context: &[String], _segment: &Segment) -> Option<String> {
        None
    }
}

#[derive(Clone)]
pub struct PerComponentStrategy {}

impl PerComponentStrategy {
    pub const fn new() -> Self {
        Self {}
    }
}

impl EntryPolicy for PerComponentStrategy {
    fn get_entry_for_sym(&self, context: &[String], segment: &Segment) -> Option<String> {
        panic!("Not implemented")
        /*
        context.first().map_or_else(
            || Some(ENTRY_SEGMENTS.to_string()),
            |root| Some([&segment.origin, "_entry_", root].concat()),
        )
        */
    }
}

#[derive(Clone)]
pub struct SmartStrategy {}

impl SmartStrategy {
    pub const fn new() -> Self {
        Self {}
    }
}

impl EntryPolicy for SmartStrategy {
    fn get_entry_for_sym(&self, context: &[String], segment: &Segment) -> Option<String> {
        panic!("Not implemented")
        /*
        // Event handlers without scope variables are put into a separate file
        if segment.scoped_idents.is_empty()
            && (segment.ctx_kind != SegmentKind::Function || &segment.ctx_name == "event$")
        {
            return None;
        }

        // Everything else is put into a single file per component
        // This means that all QRLs for a component are loaded together
        // if one is used
        context.first().map_or_else(
            // Top-level QRLs are put into a separate file
            || None,
            // Other QRLs are put into a file named after the original file + the root component
            |root| Some([&segment.origin, "_entry_", root].concat()),
        )
        */
    }
}

pub fn parse_entry_strategy(strategy: &EntryStrategy) -> Box<dyn EntryPolicy> {
    match strategy {
        EntryStrategy::Inline | EntryStrategy::Hoist => Box::<InlineStrategy>::default(),
        EntryStrategy::Hook => Box::new(PerSegmentStrategy::new()),
        EntryStrategy::Segment => Box::new(PerSegmentStrategy::new()),
        EntryStrategy::Single => Box::new(SingleStrategy::new()),
        EntryStrategy::Component => Box::new(PerComponentStrategy::new()),
        EntryStrategy::Smart => Box::new(SmartStrategy::new()),
    }
}
