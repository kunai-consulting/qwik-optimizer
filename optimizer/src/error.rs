use crate::illegal_code::IllegalCodeType;

#[derive(thiserror::Error, Debug)]
pub enum Error {
    #[error("Generic error: {0}")]
    Generic(String),

    #[error("Failed to convert OsStr, '{0}'. Context: {1}")]
    StringConversion(String, String),

    #[error("Unsupported language from SourceType: {0}")]
    UnsupportedLanguage(String),

    #[error(transparent)]
    IO(#[from] std::io::Error),

    #[error(transparent)]
    OxcUnknownExtension(#[from] oxc_span::UnknownExtension),

    #[error( "Reference to identifier '{id}' can not be used inside a Qrl($) scope because it's a {expr_type}", id =.0.identifier(), expr_type = .0.expression_type())]
    IllegalCode(IllegalCodeType),
}
