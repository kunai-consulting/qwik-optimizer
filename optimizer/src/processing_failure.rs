use crate::illegal_code::IllegalCodeType;

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone)]
pub enum ProcessingFailure {
    IllegalCode(IllegalCodeType),
}

impl From<&IllegalCodeType> for ProcessingFailure {
    fn from(value: &IllegalCodeType) -> Self {
        ProcessingFailure::IllegalCode(value.clone())
    }
}
