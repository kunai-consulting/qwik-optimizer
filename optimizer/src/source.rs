use crate::component::{Language, SourceInfo};
use crate::prelude::*;
use std::fs;
use std::path::Path;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum Source {
    ScriptFile {
        text: String,
        source_info: SourceInfo,
    },
}

impl Source {
    pub fn from_file<P: AsRef<Path>>(path_buf: P) -> Result<Self> {
        let text = fs::read_to_string(&path_buf)?;
        let source_info = SourceInfo::new(&path_buf)?;
        Ok(Source::ScriptFile { text, source_info })
    }

    pub fn from_source<T: AsRef<str>>(
        text: T,
        language: Language,
        name: Option<String>,
    ) -> Result<Self> {
        let name = name.unwrap_or("script".into());

        let path_str = format!("./{}.{}", name, language.extension());
        let path = Path::new(path_str.as_str());
        let source_info = SourceInfo::new(path)?;
        let text = text.as_ref().to_string();
        Ok(Source::ScriptFile { text, source_info })
    }

    pub fn source_code(&self) -> &str {
        match self {
            Source::ScriptFile { text, .. } => text.as_ref(),
        }
    }

    pub fn source_info(&self) -> &SourceInfo {
        match self {
            Source::ScriptFile { source_info, .. } => source_info,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const TEST_FILE: &str = "./src/test_input/test_example_1.tsx";

    #[test]
    fn can_load_from_file() {
        let source = Source::from_file(TEST_FILE).unwrap();
        let expected_source = fs::read_to_string(TEST_FILE).unwrap();
        let expected_source_info = SourceInfo::new(TEST_FILE).unwrap();

        assert_eq!(source.source_info(), &expected_source_info);
        assert_eq!(source.source_code(), &expected_source);
    }
}
