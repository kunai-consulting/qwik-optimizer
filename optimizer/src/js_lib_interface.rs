// Defines the interface between qwik-optimizer and the qwik JS library.
// This includes some types from the original qwik project that are otherwise
// not used in this rewrite, but needed for compatibility.

use crate::entry_strategy::*;
use crate::error::Error;
use crate::prelude::*;
use crate::processing_failure::ProcessingFailure;
use crate::source::Source;
use crate::transform::*;

use crate::component::*;
use serde::{Deserialize, Serialize};
use std::iter::Sum;

use std::cmp::Ordering;
use std::hash::{DefaultHasher, Hasher};
use std::path::Path;
use std::str;

#[derive(Debug, Deserialize, Copy, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum MinifyMode {
    Simplify,
    None,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TransformModuleInput {
    pub path: String,
    pub dev_path: Option<String>,
    pub code: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TransformModulesOptions {
    pub src_dir: String,
    pub root_dir: Option<String>,
    pub input: Vec<TransformModuleInput>,
    pub source_maps: bool,
    pub minify: MinifyMode,
    pub transpile_ts: bool,
    pub transpile_jsx: bool,
    pub preserve_filenames: bool,
    pub entry_strategy: EntryStrategy,
    pub explicit_extensions: bool,
    pub mode: Target,
    pub scope: Option<String>,

    pub core_module: Option<String>,
    pub strip_exports: Option<Vec<String>>,
    pub strip_ctx_name: Option<Vec<String>>,
    pub strip_event_handlers: bool,
    pub reg_ctx_name: Option<Vec<String>>,
    pub is_server: Option<bool>,
}

#[derive(Debug, Serialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct TransformOutput {
    pub modules: Vec<TransformModule>,
    pub diagnostics: Vec<Diagnostic>,
    pub is_type_script: bool,
    pub is_jsx: bool,
}

impl TransformOutput {
    pub fn append(mut self, output: &mut Self) -> Self {
        self.modules.append(&mut output.modules);
        self.diagnostics.append(&mut output.diagnostics);
        self.is_type_script = self.is_type_script || output.is_type_script;
        self.is_jsx = self.is_jsx || output.is_jsx;
        self
    }
}

impl Sum<Self> for TransformOutput {
    fn sum<I>(iter: I) -> Self
    where
        I: Iterator<Item = Self>,
    {
        iter.fold(Self::default(), |x, mut y| x.append(&mut y))
    }
}

#[derive(Debug, Serialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct TransformModule {
    pub path: String,
    pub code: String,

    pub map: Option<String>,

    pub segment: Option<SegmentAnalysis>,
    pub is_entry: bool,

    #[serde(skip_serializing)]
    pub order: u64,
}

#[derive(Debug, Clone, Copy, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum SegmentKind {
    Function,
    EventHandler,
    JSXProp,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SegmentAnalysis {
    pub origin: String,
    pub name: String,
    pub entry: Option<String>,
    pub display_name: String,
    pub hash: String,
    pub canonical_filename: String,
    pub path: String,
    pub extension: String,
    pub parent: Option<String>,
    pub ctx_kind: SegmentKind,
    pub ctx_name: String,
    pub captures: bool,
    pub loc: (u32, u32),
}

#[derive(Debug, Serialize, Clone, Eq, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SourceLocation {
    lo: usize,
    hi: usize,
    start_line: usize,
    start_col: usize,
    end_line: usize,
    end_col: usize,
}

// impl SourceLocation {
//     pub fn from(source_map: &swc_common::SourceMap, span: swc_common::Span) -> Self {
//         let start = source_map.lookup_char_pos(span.lo);
//         let end = source_map.lookup_char_pos(span.hi);
//         // - SWC's columns are exclusive, ours are inclusive (column - 1)
//         // - SWC has 0-based columns, ours are 1-based (column + 1)
//         // = +-0
//
//         Self {
//             lo: span.lo.0 as usize,
//             hi: span.hi.0 as usize,
//             start_line: start.line,
//             start_col: start.col_display + 1,
//             end_line: end.line,
//             end_col: end.col_display,
//         }
//     }
// }

impl PartialOrd for SourceLocation {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        match self.start_line.cmp(&other.start_line) {
            Ordering::Equal => self.start_col.partial_cmp(&other.start_col),
            o => Some(o),
        }
    }
}

#[derive(Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Diagnostic {
    pub category: DiagnosticCategory,
    pub code: Option<String>,
    pub file: String,
    pub message: String,
    pub highlights: Option<Vec<SourceLocation>>,
    pub suggestions: Option<Vec<String>>,
    pub scope: DiagnosticScope,
}

#[derive(Serialize, Debug, Eq, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum DiagnosticCategory {
    /// Fails the build with an error.
    Error,
    /// Logs a warning, but the build does not fail.
    Warning,
    /// An error if this is source code in the project, or a warning if in node_modules.
    SourceError,
}

#[derive(Serialize, Debug, Eq, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum DiagnosticScope {
    Optimizer,
}

fn error_to_diagnostic(error: ProcessingFailure, path: &Path) -> Diagnostic {
    let message = match error {
        ProcessingFailure::IllegalCode(code) =>
            format!(
                "Reference to identifier '{id}' can not be used inside a Qrl($) scope because it's a {expr_type}",
                id = code.identifier(), expr_type = code.expression_type()
            )
    };
    Diagnostic {
        category: DiagnosticCategory::Error,
        code: None,
        file: path.to_string_lossy().to_string(),
        message,
        highlights: None,
        suggestions: None,
        scope: DiagnosticScope::Optimizer,
    }
}

pub fn transform_modules(config: TransformModulesOptions) -> Result<TransformOutput> {
    let mut final_output = config
        .input
        .into_iter()
        .map(|input| -> Result<Option<TransformOutput>> {
            let path = Path::new(&input.path);
            let ext = path.extension().and_then(|e| e.to_str()).unwrap_or("");
            let relative_path = if path.is_relative() {
                path.into()
            } else {
                pathdiff::diff_paths(path, &config.src_dir).ok_or_else(|| {
                    Error::Generic(format!(
                        "Path {} cannot be made relative to directory {}",
                        path.to_string_lossy(),
                        &config.src_dir
                    ))
                })?
            }
            .to_string_lossy()
            .to_string();
            // TODO: Parse JSX/TSX when extension matches
            let language = match ext {
                "ts" => {
                    if config.transpile_ts {
                        Language::Typescript
                    } else {
                        return Ok(None);
                    }
                }
                "tsx" => {
                    if config.transpile_ts && config.transpile_jsx {
                        Language::Typescript
                    } else {
                        return Ok(None);
                    }
                }
                "js" => Language::Javascript,
                "jsx" => {
                    if config.transpile_jsx {
                        Language::Javascript
                    } else {
                        return Ok(None);
                    }
                }
                _ => {
                    return Ok(None);
                }
            };
            let OptimizationResult {
                optimized_app,
                errors,
            } = transform(
                Source::from_source(input.code, language, Some(relative_path.clone()))?,
                TransformOptions {
                    minify: match config.minify {
                        MinifyMode::Simplify => true,
                        MinifyMode::None => false,
                    },
                    target: config.mode,
                    transpile_jsx: config.transpile_jsx,
                },
            )?;
            let mut hasher = DefaultHasher::new();
            hasher.write(relative_path.as_bytes());
            let mut modules = vec![TransformModule {
                path: relative_path.clone(),
                code: optimized_app.body,
                map: None,
                segment: None,
                is_entry: false,
                order: hasher.finish(),
            }];
            modules.extend(
                optimized_app
                    .components
                    .into_iter()
                    .map(|c| TransformModule {
                        path: relative_path.clone(),
                        code: c.code,
                        map: None,
                        segment: Some(SegmentAnalysis {
                            origin: relative_path.clone(),
                            name: c.id.symbol_name.clone(),
                            entry: None,
                            display_name: c.id.display_name,
                            hash: c.id.hash,
                            canonical_filename: c.id.local_file_name,
                            path: relative_path.clone(),
                            extension: ext.into(),
                            parent: c.id.scope,
                            ctx_kind: if c.id.symbol_name.starts_with("on") {
                                SegmentKind::JSXProp
                            } else {
                                SegmentKind::Function
                            },
                            ctx_name: c.id.symbol_name,
                            captures: false,
                            loc: (0, 0),
                        }),
                        is_entry: true,
                        order: c.id.sort_order,
                    }),
            );
            Ok(Some(TransformOutput {
                modules,
                diagnostics: errors
                    .into_iter()
                    .map(|e| error_to_diagnostic(e, &path))
                    .collect(),
                is_type_script: config.transpile_ts, // TODO: Set this flag correctly
                is_jsx: config.transpile_jsx,        // TODO: Set this flag correctly
            }))
        })
        .sum::<Result<Option<TransformOutput>>>()?
        .unwrap_or(TransformOutput::default());

    final_output.modules.sort_unstable_by_key(|key| key.order);

    Ok(final_output)
}

#[cfg(test)]
mod tests {
    use super::*;
    use glob::glob;
    use std::path::PathBuf;

    #[test]
    fn test_project_1() {
        // This should be a macro eventually
        let func_name = function_name!();
        let path = PathBuf::from("./src/test_input").join(func_name);

        println!(
            "Loading test input project directory from path: {:?}",
            &path
        );

        let result = transform_modules(TransformModulesOptions {
            input: glob(path.join("src/**/*.ts*").to_str().unwrap())
                .unwrap()
                .into_iter()
                .map(|file| {
                    let file = Path::new(".").join(file.unwrap());
                    let code = std::fs::read_to_string(&file).unwrap();
                    TransformModuleInput {
                        path: file.into_os_string().into_string().unwrap(),
                        dev_path: None,
                        code,
                    }
                })
                .collect(),
            src_dir: path.clone().into_os_string().into_string().unwrap(),
            root_dir: Some(path.clone().into_os_string().into_string().unwrap()),
            minify: MinifyMode::None,
            entry_strategy: EntryStrategy::Component,
            source_maps: false,
            transpile_ts: true,
            transpile_jsx: true,
            preserve_filenames: false,
            explicit_extensions: false,
            mode: Target::Dev,
            scope: None,

            core_module: None,
            strip_exports: None,
            strip_ctx_name: None,
            strip_event_handlers: false,
            reg_ctx_name: None,
            is_server: None,
        })
        .unwrap();

        insta::assert_yaml_snapshot!(func_name, result);
    }
}
