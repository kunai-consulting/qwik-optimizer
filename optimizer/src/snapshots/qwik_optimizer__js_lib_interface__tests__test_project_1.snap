---
source: src/js_lib_interface.rs
expression: result
---
modules:
  - path: src/routes/index.tsx
    code: "import { qrl } from \"@qwik.dev/core\";\nimport { componentQrl } from \"@qwik.dev/core\";\nimport { DocumentHead } from \"@qwik.dev/router\";\nexport default componentQrl(qrl(() => import(\"./src/routes/index.tsx.tsx_component_0lHKhlBXq9w\"), \"component_0lHKhlBXq9w\"));\nexport const head: DocumentHead = componentQrl(qrl(() => import(\"./src/routes/index.tsx.tsx_component_0lHKhlBXq9w\"), \"component_0lHKhlBXq9w\"));\n"
    map: ~
    segment: ~
    isEntry: false
  - path: src/root.tsx
    code: "import { QwikCityProvider } from \"@qwik.dev/router\";\nimport { RouterHead } from \"./components/router-head/router-head\";\nimport { RouterOutlet } from \"@qwik.dev/router\";\nimport { ServiceWorkerRegister } from \"@qwik.dev/router\";\nimport { isDev } from \"@qwik.dev/core\";\nexport const component_Q1jqh14YGE4 = () => {\n\treturn <QwikCityProvider>\n      <head>\n        <meta charset=\"utf-8\" />\n        {!isDev && <link rel=\"manifest\" href={`${import.meta.env.BASE_URL}manifest.json`} />}\n        <RouterHead />\n      </head>\n      <body lang=\"en\">\n        <RouterOutlet />\n        {!isDev && <ServiceWorkerRegister />}\n      </body>\n    </QwikCityProvider>;\n};\n"
    map: ~
    segment: ~
    isEntry: true
  - path: src/components/router-head/router-head.tsx
    code: "import { qrl } from \"@qwik.dev/core\";\nimport { componentQrl } from \"@qwik.dev/core\";\n/**\n* The RouterHead component is placed inside of the document `<head>` element.\n*/\nexport const RouterHead = componentQrl(qrl(() => import(\"./src/components/router-head/router-head.tsx.tsx_RouterHead_component_cmnVjLRycb8\"), \"RouterHead_component_cmnVjLRycb8\"));\n"
    map: ~
    segment: ~
    isEntry: false
  - path: src/root.tsx
    code: "import { qrl } from \"@qwik.dev/core\";\nimport { componentQrl } from \"@qwik.dev/core\";\nexport default componentQrl(qrl(() => import(\"./src/root.tsx.tsx_component_Q1jqh14YGE4\"), \"component_Q1jqh14YGE4\"));\n"
    map: ~
    segment: ~
    isEntry: false
  - path: src/routes/service-worker.ts
    code: "import { setupServiceWorker } from \"@qwik.dev/router/service-worker\";\nsetupServiceWorker();\naddEventListener(\"install\", () => self.skipWaiting());\naddEventListener(\"activate\", () => self.clients.claim());\ndeclare const self: ServiceWorkerGlobalScope;\n"
    map: ~
    segment: ~
    isEntry: false
  - path: src/components/router-head/router-head.tsx
    code: "import { useDocumentHead } from \"@qwik.dev/router\";\nimport { useLocation } from \"@qwik.dev/router\";\nexport const RouterHead_component_cmnVjLRycb8 = () => {\n\tconst head = useDocumentHead();\n\tconst loc = useLocation();\n\treturn <>\n      <title>{head.title}</title>\n\n      <link rel=\"canonical\" href={loc.url.href} />\n      <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\" />\n      <link rel=\"icon\" type=\"image/svg+xml\" href=\"/favicon.svg\" />\n\n      {head.meta.map((m) => <meta key={m.key} {...m} />)}\n\n      {head.links.map((l) => <link key={l.key} {...l} />)}\n\n      {head.styles.map((s) => <style key={s.key} {...s.props} {...s.props?.dangerouslySetInnerHTML ? {} : { dangerouslySetInnerHTML: s.style }} />)}\n\n      {head.scripts.map((s) => <script key={s.key} {...s.props} {...s.props?.dangerouslySetInnerHTML ? {} : { dangerouslySetInnerHTML: s.script }} />)}\n    </>;\n};\n"
    map: ~
    segment: ~
    isEntry: true
  - path: src/routes/layout.tsx
    code: "import { Slot } from \"@qwik.dev/core\";\nexport const component_9FUyuTlHyso = () => {\n\treturn <Slot />;\n};\n"
    map: ~
    segment: ~
    isEntry: true
  - path: src/entry.dev.tsx
    code: "import Root from \"./root\";\nimport { render } from \"@qwik.dev/core\";\nimport { RenderOptions } from \"@qwik.dev/core\";\nexport default function(opts: RenderOptions) {\n\treturn render(document, <Root />, opts);\n}\n"
    map: ~
    segment: ~
    isEntry: false
  - path: src/entry.ssr.tsx
    code: "import Root from \"./root\";\nimport { renderToStream } from \"@qwik.dev/core/server\";\nimport { manifest } from \"@qwik-client-manifest\";\nimport { RenderToStreamOptions } from \"@qwik.dev/core/server\";\nexport default function(opts: RenderToStreamOptions) {\n\treturn renderToStream(<Root />, {\n\t\tmanifest,\n\t\t...opts,\n\t\tcontainerAttributes: {\n\t\t\tlang: \"en-us\",\n\t\t\t...opts.containerAttributes\n\t\t},\n\t\tserverData: { ...opts.serverData }\n\t});\n}\n"
    map: ~
    segment: ~
    isEntry: false
  - path: src/routes/index.tsx
    code: "export const component_0lHKhlBXq9w = () => {\n\treturn <>\n      <h1>Hi ðŸ‘‹</h1>\n      <div>\n        Can't wait to see what you build with qwik!\n        <br />\n        Happy coding.\n      </div>\n    </>;\n};\n"
    map: ~
    segment: ~
    isEntry: true
  - path: src/entry.preview.tsx
    code: "import render from \"./entry.ssr\";\nimport qwikCityPlan from \"@qwik-city-plan\";\nimport { createQwikCity } from \"@qwik.dev/router/middleware/node\";\n/**\n* The default export is the QwikCity adapter used by Vite preview.\n*/\nexport default createQwikCity({\n\trender,\n\tqwikCityPlan\n});\n"
    map: ~
    segment: ~
    isEntry: false
  - path: src/routes/layout.tsx
    code: "import { qrl } from \"@qwik.dev/core\";\nimport { componentQrl } from \"@qwik.dev/core\";\nimport { RequestHandler } from \"@qwik.dev/router\";\nexport const onGet: RequestHandler = async ({ cacheControl }) => {\n\tcacheControl({\n\t\tstaleWhileRevalidate: 60 * 60 * 24 * 7,\n\t\tmaxAge: 5\n\t});\n};\nexport default componentQrl(qrl(() => import(\"./src/routes/layout.tsx.tsx_component_9FUyuTlHyso\"), \"component_9FUyuTlHyso\"));\n"
    map: ~
    segment: ~
    isEntry: false
diagnostics: []
isTypeScript: true
isJsx: true
